File: Examples/Objective-C/ExampleApp/TaboolaDemoApp/TBFeedCollectionView.m
//
//  ViewController.m
//  UICollectionViewAppWithTaboola
//
//  Created by Yaromyr Oleksyshun on 5/25/18.
//  Copyright © 2018 Vakoms. All rights reserved.
//

#import "TBFeedCollectionView.h"
#import "TaboolaCollectionViewCell.h"

@interface TBFeedCollectionView () <TaboolaViewDelegate>

#pragma mark - Properties

@property (nonatomic) TaboolaCollectionViewCell* taboolaCell;
@end

@implementation TBFeedCollectionView {
    NSUInteger taboolaCellIndex;
}


#pragma mark - ViewController lifecycle
- (void)viewDidLoad {
    [super viewDidLoad];
    taboolaCellIndex = 1;
}


#pragma mark - UICollectionViewDatasource
- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
    return 2;
}

- (NSInteger)numberOfSectionsInCollectionView:(UICollectionView *)collectionView {
    return 1;
}

-(CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout sizeForItemAtIndexPath:(NSIndexPath *)indexPath {
    return indexPath.item == taboolaCellIndex ? CGSizeMake(self.view.frame.size.width, TaboolaView.widgetHeight) : CGSizeMake(self.view.frame.size.width, 200);
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath {
    
    if ([indexPath item] == taboolaCellIndex) {
        if (!self.taboolaCell) {
            self.taboolaCell = [collectionView dequeueReusableCellWithReuseIdentifier:@"TaboolaCell" forIndexPath:indexPath];
            self.taboolaCell.taboolaView.delegate = self;
            self.taboolaCell.taboolaView.mode = @"thumbnails-feed";
            self.taboolaCell.taboolaView.publisher = @"betterbytheminute-app";
            self.taboolaCell.taboolaView.pageType = @"article";
            self.taboolaCell.taboolaView.pageUrl = @"http://www.example.com";
            self.taboolaCell.taboolaView.placement = @"feed-sample-app";
            self.taboolaCell.taboolaView.targetType = @"mix";
            self.taboolaCell.taboolaView.logLevel = LogLevelDebug;
            [self.taboolaCell.taboolaView setInterceptScroll:YES];
            [self.taboolaCell.taboolaView fetchContent];
        }
        return self.taboolaCell;
    } else {
        UICollectionViewCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:@"randomCell" forIndexPath:indexPath];
        return cell;
    }
}

#pragma mark - TaboolaView delegate
-(void)taboolaView:(UIView *)taboolaView didLoadPlacementNamed:(NSString *)placementName withHeight:(CGFloat)height {
    
}

-(void)taboolaView:(UIView *)taboolaView placementNamed:(NSString *)placementName resizedToHeight:(CGFloat)height {

}

- (void)taboolaView:(UIView *)taboolaView didFailToLoadPlacementNamed:(NSString *)placementName withErrorMessage:(NSString *)error {
    
}

- (BOOL)onItemClick:(NSString *)placementName withItemId:(NSString *)itemId withClickUrl:(NSString *)clickUrl isOrganic:(BOOL)organic {
    return YES;
}

@end

File: Examples/Swift/TaboolaDemoSwiftApp/TaboolaDemoSwiftApp/TBFeedCollectionView.swift
//
//  TBFeedCollectionView.swift
//  TaboolaDemoSwiftApp
//
//  Created by Tzufit Lifshitz on 6/19/18.
//  Copyright © 2018 Taboola. All rights reserved.
//

import UIKit
import TaboolaSDK

private let taboolaCellIdentifier = "TaboolaCell"
private let randomCellIdentifier = "randomCell"

class TaboolaCollectionViewCell : UICollectionViewCell {
    @IBOutlet weak var taboolaView: TaboolaView!
}

class TBFeedCollectionView: UICollectionViewController , TaboolaViewDelegate {
    
    var taboolaSection = 1
    var taboolaCell:TaboolaCollectionViewCell?

    override func viewDidLoad() {
        super.viewDidLoad()

        // Uncomment the following line to preserve selection between presentations
        // self.clearsSelectionOnViewWillAppear = false

        // Register cell classes
//        self.collectionView!.register(UICollectionViewCell.self, forCellWithReuseIdentifier: taboolaCellIdentifier)

        // Do any additional setup after loading the view.
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

    /*
    // MARK: - Navigation

    // In a storyboard-based application, you will often want to do a little preparation before navigation
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        // Get the new view controller using [segue destinationViewController].
        // Pass the selected object to the new view controller.
    }
    */

    // MARK: UICollectionViewDataSource

    override func numberOfSections(in collectionView: UICollectionView) -> Int {
        // #warning Incomplete implementation, return the number of sections
        return 2
    }


    override func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        // #warning Incomplete implementation, return the number of items
        return 1
    }

    override func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        
        if indexPath.section == taboolaSection {
            if  taboolaCell == nil {
                taboolaCell = collectionView.dequeueReusableCell(withReuseIdentifier: taboolaCellIdentifier, for: indexPath) as? TaboolaCollectionViewCell
                taboolaCell!.taboolaView.delegate = self
                taboolaCell!.taboolaView.mode = "thumbnails-feed"
                taboolaCell!.taboolaView.publisher = "betterbytheminute-app"
                taboolaCell!.taboolaView.pageType = "article"
                taboolaCell!.taboolaView.pageUrl = "http://www.example.com"
                taboolaCell!.taboolaView.placement = "feed-sample-app"
                taboolaCell!.taboolaView.targetType = "mix"
                taboolaCell!.taboolaView.enableClickHandler = true
                taboolaCell!.taboolaView.setInterceptScroll(true)
                taboolaCell!.taboolaView.fetchContent()
            }
            return taboolaCell!
        } else {
            let cell = collectionView.dequeueReusableCell(withReuseIdentifier: randomCellIdentifier, for: indexPath)
            return cell
        }
    }

}

extension TBFeedCollectionView: ActionAssistantProtocol {
    func refreshChild() {
        //        mTaboolaView.refresh()
    }
    
    func editChild(dict: [String : String]) {
        taboolaCell!.taboolaView.mode = dict[ConstantsProperties.mode]
        taboolaCell!.taboolaView.publisher = dict[ConstantsProperties.publisher]
        taboolaCell!.taboolaView.pageType = dict[ConstantsProperties.pageType]
        taboolaCell!.taboolaView.pageUrl = dict[ConstantsProperties.pageUrl]
        taboolaCell!.taboolaView.placement = dict[ConstantsProperties.placement]
        taboolaCell!.taboolaView.fetchContent()
        self.collectionView?.reloadData()
    }
}
